Contact Info & Map Widget: Remove Geocoding and add map to form

kbrown9 opened the pull request:
Fixes #10355
Fixes #12925 
Fixes #12913 

#### Changes proposed in this Pull Request:
* Remove the use of the Google Maps Geocoding API. The Geocoding API is only used to verify that the address is valid. After these changes are made, users will need to enable only one Google Maps API (Maps Embed API).

* Add the Google Maps embedded map to the widget's admin form. This change will allow users to view the maps that will be shown on their site in the admin form. The users will be able to verify that the maps are correct and receive a useful error if the API key is invalid.

#### Is this a new feature or does it add/remove features to an existing part of Jetpack?
* This change adds a feature to the existing Contact Info & Map widget.

#### Testing instructions:
1) Add a new Contact Info & Maps widget (Appearance -> Widgets -> Contact Info & Map), or access an existing Contact Info & Maps widget.

2) Check the Show Map checkbox to show the API key box and the embedded map.

3) The embedded map iframe will show this error when an API key is not entered:
     > Google Maps Platform rejected your request. You must use an API key to authenticate each request to Google Maps Platform APIs. For additional information, please refer to http://g.co/dev/maps-no-account

4) Enter an invalid API key. The embedded map iframe will show this error when an invalid API key is entered:
     > Google Maps Platform rejected your request. The provided API key is invalid.

5) Enter a valid Google Maps API key. (The Embed Maps API must be enabled for the API key.) Save the widget, and the embedded map will be shown in the admin form.

#### Proposed changelog entry for your changes:
Contact Info Widget: Remove use of the Geocoding API and add an embedded map to the admin form

kbrown9 committed:
Contact Info & Map Widget: Remove Geocoding and add map to form

The Google Maps Geocoding API is used to verify that the given address
is valid. However, use of this API has recently caused problems for
users. Users need to enable two Google Maps APIs in order to use this
widget, which isn't user-friendly.

Instead of using the Geocoding API results to verify the address, show
the embedded map in the widget's admin form. The user can verify that the
address is valid and correct by looking at the map. Also, the user will
receive Google Maps' error messages if the API key is not valid.

Do not show the embedded map In the Customizer admin form. The embedded
map isn't necessary in this form because the embedded map is shown in
the Customizer preview.

kbrown9 commented:
I just noticed that the Customize page will show inconsistent maps when an address is changed. The preview page will show the map for the new address. However, the map in the widget's admin form doesn't update, so it shows the map for the old address. 

I'll try to think of a solution for this.

kbrown9 commented:
Following-up on my last comment - I addressed the issue with inconsistent maps on the Customizer page in the latest commit. In this commit, the embedded map is not displayed in the widget admin form on the Customizer page. Since the Customizer preview displays the embedded map, it's not necessary to display the map in the widget admin form. 

I used the `is_customize_preview()` function ([code reference](https://developer.wordpress.org/reference/functions/is_customize_preview/)) to detect whether the widget admin form is being displayed on the Customizer page.

jeherve committed:
Rely on esc_url_raw for remote call

https://codex.wordpress.org/Function_Reference/esc_url_raw

jeherve committed:
Clarify update process for the map, and coding standard changes

- A few additional comments help understand when has_good_map is needed.
- Use Yoda conditions, we must

jeherve committed:
Simplify addition of query params to Google embed URL.

jeherve committed:
Only attempt to display the map when we have info to display it

jeherve committed:
Enqueue CSS on widgets page as well

This avoids that the map breaks out of the widget area layout when displayed.

jeherve committed:
Display output from API request in widget area when error

- This gives more information to users who entered an API key that is not correct for example, instead of just not displaying any map at all.

jeherve reviewed (changes requested):
Excellent work! I'd like to suggest a few changes, and to make things easier I've just submitted this PR to your branch:
https://github.com/kbrown9/jetpack/pull/1

If you are happy with those changes, you can merge the PR and my changes will be automatically added to your PR here.

Beyond those changes, I think we should take that opportunity to better display error messages in the customizer when something is wrong. 
Those could be displayed in the widget settings directly, just like on the old settings screen. They could also be displayed in the widget preview directly, since that's where site owners will check what the map will look like.

Let me know what you think.

kbrown9 commented:
> 
> 
> Excellent work! I'd like to suggest a few changes, and to make things easier I've just submitted this PR to your branch:
> [kbrown9#1](https://github.com/kbrown9/jetpack/pull/1)
> 
> If you are happy with those changes, you can merge the PR and my changes will be automatically added to your PR here.
> 
> Beyond those changes, I think we should take that opportunity to better display error messages in the customizer when something is wrong.
> Those could be displayed in the widget settings directly, just like on the old settings screen. They could also be displayed in the widget preview directly, since that's where site owners will check what the map will look like.
> 
> Let me know what you think.

Thank you for this review! I think all of these changes look great.

I'll work on displaying the error message in the customizer!

kbrown9 committed:
Improve code readability and remove an API call.

 * Move the logic for determining whether to update the instance's
   'goodmap' value from the 'update()' function to a new function,
   'update_goodmap()'.

 * Instead of calling 'has_good_map()' in the 'form()' function, just
   check the instance's 'goodmap' value. We've already called
   'has_good_map()' and saved the result in the instance's 'goodmap'
   value, so no need to call "has_good_map()" again.

kbrown9 committed:
Show the API key error in the widget in the customizer preview

When the Google Maps API returns an error, display it in the widget in
the Customizer preview. The CSS properties for the
'contact-map-api-error' class are intended to match the CSS properties
for the admin 'notice-warning' class.

jeherve commented:
@kbrown9 Is this ready for another review?

kbrown9 commented:
> 
> 
> @kbrown9 Is this ready for another review?

Yes, it's ready now! I just pushed my last commit. So, I've added three new commits on top of your changes:

commit [7f4b2e09](https://github.com/Automattic/jetpack/pull/12474/commits/7f4b2e09a0be61cd61344dcb2bbe29627438fa1b): This commit includes changes that I think improve readability.

commit [43cb2133](https://github.com/Automattic/jetpack/pull/12474/commits/43cb2133c55ed71d9240c317eec4fe5b63b070b8): This commit displays API key errors in the Contact Info widgets on the Customizer preview page.

commit [15910c91](https://github.com/Automattic/jetpack/pull/12474/commits/15910c911c7495a99dd587b6ee89ee1f71724cbd): This commit displays API key errors in the Contact Info admin forms on the Customizer preview page.

Please let me know if you'd prefer a different approach for these error messages!

Finally, what's the workflow for handling the commits that are improvements to previous commits? After the review is complete, should the commits be squashed together?

jeherve commented:
> Yes, it's ready now!

Excellent, I'll take a look when I can.

> what's the workflow for handling the commits that are improvements to previous commits? After the review is complete, should the commits be squashed together?

Don't worry about it :) I'll squash all the commits when I merge to `master`.

jeherve reviewed (changes requested):
It looks like Unit Tests are failing right now:

```
modules/widgets/contact-info/contact-info-admin.js
  line 23  col 9   'const' is available in ES6 (use 'esversion: 6') or Mozilla JS extensions (use moz).
  line 29  col 13  'const' is available in ES6 (use 'esversion: 6') or Mozilla JS extensions (use moz).
  line 30  col 13  'const' is available in ES6 (use 'esversion: 6') or Mozilla JS extensions (use moz).
  line 25  col 17  'ajaxurl' is not defined.
  ⚠  4 warnings
```

The first 3 warnings are an annoying consequence of us still using jshint (see #11915), but I'm afraid we won't be able to merge as long as the tests are failing.

kbrown9 commented:
> 
> 
> It looks like Unit Tests are failing right now:
> 
> ```
> modules/widgets/contact-info/contact-info-admin.js
>   line 23  col 9   'const' is available in ES6 (use 'esversion: 6') or Mozilla JS extensions (use moz).
>   line 29  col 13  'const' is available in ES6 (use 'esversion: 6') or Mozilla JS extensions (use moz).
>   line 30  col 13  'const' is available in ES6 (use 'esversion: 6') or Mozilla JS extensions (use moz).
>   line 25  col 17  'ajaxurl' is not defined.
>   ⚠  4 warnings
> ```
> 
> The first 3 warnings are an annoying consequence of us still using jshint (see #11915), but I'm afraid we won't be able to merge as long as the tests are failing.

Thanks for the info! I'll get these fixed up.

kbrown9 committed:
Contact Info & Map Widget: Dynamically update customizer admin form

Display API key errors in the Contact Info & Map widget's admin form on
the Customizer preview page. This requres an Ajax call, which results in
a call to the "has_good_map()" method. The "has_good_map()" method
checks the API key by calling the Google Maps API. If the API returns an
error, that error is displayed in the admin form on the Customizer page.

kbrown9 commented:
@jeherve - I amended the commit that caused the jshint test failures. The commit passes the tests now!

jeherve committed:
Contact Info: fix linting warnings in js file

jeherve reviewed (dismissed):
I rebased to fix the last linting errors now that we got rid of jshint. It tests well, and I think this would be ready to merge. 

Leaving it to another Crew member to do a final review on this, and then we can merge!

brbrr reviewed (dismissed):
It works well in my tests. Thank you!

I noticed a few minor issues, but I don't think they should block merging of this one:

- Empty "warning" formatting is shown once "Show map" is checked (in both custumizer & `wp-admin/widgets.php`)
<img width="520" alt="Widgets ‹ HelloWord — WordPress 2019-07-26 11-59-22" src="https://user-images.githubusercontent.com/5654161/61940663-928e3e80-af9e-11e9-8035-d0bb9836465d.png">
 
- Map width is way bigger than widget box in `wp-admin/widgets.php`
<img width="920" alt="Widgets ‹ HelloWord — WordPress 2019-07-26 12-07-12" src="https://user-images.githubusercontent.com/5654161/61940767-cbc6ae80-af9e-11e9-89fe-f7fba281ee7e.png">

jeherve committed:
Contact Info CSS: update version so cache can be busted

Addresses https://github.com/Automattic/jetpack/pull/12474#pullrequestreview-267077304

jeherve committed:
Do not output notice container when no map has been set up yet

Addresses https://github.com/Automattic/jetpack/pull/12474#pullrequestreview-267077304

jeherve commented:
@brbrr I just pushed some changes that should solve those issues; do you think you could test again when you get the chance?

jeherve commented on modules/widgets/contact-info.php:
```diff
@@ -379,16 +349,85 @@ function urlencode_address( $address ) {
+		 * the Customizer preview.
+		 */
+		function ajax_check_api_key() {
+			if ( isset( $_POST['apikey'] ) ) {
```
Coming back to this, I think we should create a nonce, use it with our requests, check against them with `check_ajax_referer`, and check for `current_user_can( 'customize' )` as well.

If we don't protect access to those requests, we'll run the risk of having non logged-in users making requests instead of the admin.

kbrown9 replied:
Good point. I'll add these checks.

jeherve reviewed (changes requested):
I think we'll need to review the requests we make and make sure they are well protected and can't be abused by non-logged in users.

kbrown9 committed:
Contact Info & Map Widget: Protect the Ajax request

Use a nonce to verify that the Ajax request is valid. Also check that
the user is allowed to customize before checking the API key from the
Ajax request.

jeherve reviewed (dismissed):
Everything seems to work well for me now. 🚢 👍

mdawaffe commented:
I still see the too-wide map sometimes:

1. Add the widget.
2. Enter an invalid API key.
3. Save.
4. Do a full page refresh.
5. Enter a valid API key.
6. Save.
7. See the too-wide map.

<img width="713" alt="Screen Shot 2019-08-15 at 4 08 57 PM" src="https://user-images.githubusercontent.com/125994/63132789-3607b900-bf77-11e9-9489-0312a86ebd7f.png">

kbrown9 commented:
> I still see the too-wide map sometimes:
> 

Thanks - I was able to reproduce the too-wide map with your instructions. I'll look into this.

kbrown9 committed:
Fix stylesheet enqueue problem

Enqueue the contact info stylesheet on widgets.php using the
'admin_enqueue_scripts' hook since widgets.php is an admin page.

kbrown9 commented:
Commit c8544a7 fixes the too-wide map problem by using the `admin_enqueue_scripts` hook to enqueue the stylesheet on widgets.php.

mdawaffe commented:
Nice! @kbrown9, could you please rebase this PR on top of the latest master? That will get those failing Travis CI tests to pass.
